[#9] 질문 기능 구현

puallab opened the pull request:
## PR 유형

- [ ] 버그 수정
- [x] 기능
- [ ] 코드 스타일
- [ ] 리팩토링
- [ ] 빌드
- [ ] CI/CD
- [ ] 문서
- [x] 기타
  <br>

## 변경사항

### 변경 전
qeustion 등록 기능 X
### 변경 후
question 등록 기능 추가
schedule repository 및 인터페이스 추가

## 기타
스케줄 예약시 디렉터가 허용한 시간에만 약속 날짜를 잡을 수 있기 때문에 스케쥴 테이블에서 해당 시간이 null 이거나 closed 상태라면 exception을 발생하도록 구현하였습니다.
merge conflit 발생으로 인해 수정했습니다.

puallab committed:
#9 질문 저장 기능 구현

puallab committed:
#9 스케줄 시작 시간 검증

puallab committed:
global 예외 처리

puallab committed:
#9 confilit merge

tilsong commented on src/main/java/com/directors/domain/schedule/ScheduleStatus.java:
```diff
@@ -0,0 +1,14 @@
+package com.directors.domain.schedule;
+
+public enum ScheduleStatus {
+	OPENED("opend"), CLOSED("closed");
```
opend 오타 있습니다ㅎ

puallab replied:
수정했습니다~!

f-lab-fi commented on src/main/java/com/directors/application/question/QuestionService.java:
```diff
@@ -1,36 +1,70 @@
+
+import org.springframework.stereotype.Service;
+
+import com.directors.domain.question.Question;
+import com.directors.domain.question.QuestionRepository;
+import com.directors.domain.question.QuestionStatus;
+import com.directors.domain.schedule.Schedule;
+import com.directors.domain.schedule.ScheduleRepository;
+import com.directors.domain.schedule.ScheduleStatus;
+import com.directors.domain.user.UserRepository;
+import com.directors.infrastructure.exception.schedule.ClosedScheduleException;
+import com.directors.infrastructure.exception.schedule.InvalidMeetingTimeException;
+import com.directors.presentation.qeustion.request.CreateQuestionRequest;
+import com.directors.presentation.qeustion.response.ReceivedQuestionResponse;
+import com.directors.presentation.qeustion.response.SentQuestionResponse;
+
+import jakarta.transaction.Transactional;
+import lombok.RequiredArgsConstructor;
+
+@Service
+@RequiredArgsConstructor
+public class QuestionService {
+	private final QuestionRepository questionRepository;
+	private final UserRepository userRepository;
+	private final ScheduleRepository scheduleRepository;
+
+	public List<SentQuestionResponse> getSendList(String questionerID) {
+		List<Question> sentQuestions = questionRepository.findByQuestionerId(questionerID);
+		return sentQuestions.stream()
+			.map(question -> SentQuestionResponse.from(question))
+			.toList();
+	}
+
+	public List<ReceivedQuestionResponse> getReceiveList(String directorId) {
+		List<Question> receivedQuestions = questionRepository.findByDirectorId(directorId);
+		return receivedQuestions.stream()
+			.filter(question -> question.getStatus() != QuestionStatus.COMPLETE)
+			.map(question -> ReceivedQuestionResponse.from(question))
+			.toList();
+	}
```
여기까지 코드는 왜 지워졌다가 새로 추가된 것처럼 노출되는지 잘 모르겠군요;;
뭔가 짐작되시는게 있나용

puallab replied:
merge conflit 해결 후 커밋 메시지 수정하기 위해 rebase -i head~2를 하면서 뭔가 꼬이는 느낌이 들었는데 그 때 발생한 이유이지 않을까 싶습니다..  결국엔 reset hard로 commit 돌려놓고 했는데도 잘 처리가 안된거같네요..

f-lab-fi commented on src/main/java/com/directors/application/question/QuestionService.java:
```diff
@@ -1,36 +1,70 @@
+	private void validateTime(LocalDateTime startTime, String userId) {
+		Optional<Schedule> optionalSchedule = scheduleRepository.findByStartTimeAndUserId(startTime, userId);
+
+		if (optionalSchedule.orElseThrow(() -> new InvalidMeetingTimeException(userId))
+			.getStatus() != ScheduleStatus.OPENED) {
```
지금 이 코드 가독성이 좋지 않습니다.
if문 안쪽에 있는 코드를 발췌하면 이렇게 되는데, 이 코드가 제 3자 입장에서 잘 읽힐 수 있을까요?
그리고 혹시 더 잘 읽힐 수 있는 방법은 어떤게 있을까요.

```java
optionalSchedule.orElseThrow(() -> new InvalidMeetingTimeException(userId))
			.getStatus() != ScheduleStatus.OPENED
```

puallab replied:
안그래도 이 부분에 대해 고민을 했었습니다.
optional.isEmpty() 메서드를 사용해서 2개의 if문을 구분하여 각각 처리하는 방법과 고민했었는데 멘토님 말씀처럼 각각의 에외를 별도로 구분해서  처리하는 것이 가독성이 더 좋을 것 같습니다

f-lab-fi replied:
@puallab 

넵 if문 안에는 뭔가 식을 넣지 않는게 좋습니다. 가능하면 값이 들어오도록 만들어주세요.

f-lab-fi commented on src/main/java/com/directors/domain/question/Question.java:
```diff
@@ -1,26 +1,44 @@
+package com.directors.domain.question;
+
+import java.time.LocalDateTime;
+
```
여기도 왜 새로운 코드 베이스처럼 보이는지 ㅋ..

puallab replied:
위와 동일한 이유 같습니다...

f-lab-fi commented on src/main/java/com/directors/domain/question/Question.java:
```diff
@@ -1,26 +1,44 @@
+	private String category; // 카테고리 결정되면 enum으로 변경 예정
+	private Long scheduledId;
+	private LocalDateTime startTime;
+
+	public static Question of(CreateQuestionRequest createRequest, String questionerId, Long scheduledId) {
+		return Question.builder()
```
CreateQuestionRequest는 표현 계층과 밀접하게 관련된 클래스입니다. 도메인 계층이 표현 계층에 의존하고 있는 상태인데요, 이 의존 방향은 가급적 피하시는게 좋습니다. :)
코드를 어떻게 바꿔볼지 고민해보시죠~!

puallab replied:
of 메소드는 스태틱 팩토리 메소드로 객체를 생성하는 메서드입니다.
따라서 특정 계층에 의존하지 않고 Question 클래스의 멤버로만 이루어진 파라미터를 갖는 것으로 변경하였습니다.

f-lab-fi replied:
@puallab 

넵 좋습니다. 아니면 CreateQuestionRequest 쪽에 정적 팩토리 메서드를 만들어줘서 CreateQuestionRequest -> Question 쪽으로 의존하도록 만들어도 됩니다. 표현 계층이 도메인 계층에 의존하는건 자연스러운 의존 방향입니다.

f-lab-fi commented on src/main/java/com/directors/domain/schedule/ScheduleStatus.java:
```diff
@@ -0,0 +1,14 @@
+package com.directors.domain.schedule;
+
+public enum ScheduleStatus {
+	OPENED("opend"), CLOSED("closed");
+	private final String value;
```
opend <-- 오타같습니다.

puallab replied:
수정하겠습니다!

f-lab-fi commented on src/main/java/com/directors/infrastructure/exception/schedule/InvalidMeetingTimeException.java:
```diff
@@ -0,0 +1,17 @@
+import lombok.Getter;
+
+@Getter
+public class InvalidMeetingTimeException extends RuntimeException {
+	private String userId;
+
+	public InvalidMeetingTimeException() {
+	}
+
+	public InvalidMeetingTimeException(String userId) {
+		super("디렉터가 허용한 만남시간이 아닙니다.");
+		this.userId = userId;
+	}
```
[InvalidMeetingTimeException가 던져지는 상황](https://github.com/f-lab-edu/directors/pull/13/files#diff-5316fce7ccf7b0550667bdc5e01146f5432615dfd348bc8e750fb8d5b25e3430R61-R65)은 userId 외에도 요청한 만남 시간에 대한 정보도 함께 있는게 좋지 않을까요~?

puallab replied:
넵..! 수정하겠습니다!
근데 한 가지 궁금한 점이 있습니다.
작성한 schedule 관련 예외인 InvalidMeetingTimeException과 ClosedScheduleException의 경우에 message만 빼면 중복된 코드를 갖는 예외 클래스입니다.
이런 경우에도 구체적이고 명학환 예외를 던지기 위해서 중복되더라도 매번 새로운 런타임 예외를 작성하는 것이 좋을까요?

f-lab-fi replied:
이런 상황에선 두 상황을 아우를 수 있는 이름을 가진 하나의 예외(A라고 하겠습니다)를 만들어서 사용하는게 좋습니다.
그러다가 예외의 타입을 구분하여 처리해야하는 상황이 생길 수 있습니다.
그러면 B, C라는 예외를 추가하고, B와 C는 A 예외를 상속하도록 만들면 됩니다.
이렇게 만들면 B, C는 A로 한번에 처리할 수도 있고 필요한 경우 B, C로 구체적으로 나눠서 처리할 수 있게됩니다.

결과적으로 우선은 지금 나눠서 처리할 필요가 없다면 하나의 예외로 만들어서 처리하는게 좋을 것 같다는 겁니다. :)

f-lab-fi commented on src/main/java/com/directors/infrastructure/jpa/schedule/InmemoryScheduleRepository.java:
```diff
@@ -0,0 +1,37 @@
+		}
+	};
+
+	@Override
+	public Optional<Schedule> findByStartTimeAndUserId(LocalDateTime startTime, String userId) {
+		Optional<Schedule> optionalSchedule = map.values()
+			.stream()
+			.filter(schedule -> (schedule.getUserId().equals(userId) && schedule.getStartTime().isEqual(startTime)))
+			.findFirst();
+		return optionalSchedule;
+	}
```
.filter(schedule -> (schedule.getUserId().equals(userId) && schedule.getStartTime().isEqual(startTime)))

여기서 schedule의 getter를 이용해 userId와 startTime을 가져오고 있는데요, 이 두가지 값을 가져오는 행위가 비즈니스적인 의미를 가지는건 아닐까요? (저도 명확한건 아니라서 여쭤보는겁니다. ㅎㅎ)

puallab replied:
해당 메서드는 "유저의 스케줄이 예약 가능 상태인지 확인"이란 행위를 처리하기 위한 데이터를 조회할 때 호출됩니다.
따라서 해당 코드는 비즈니스 로직을 수행하기 위해 DB에서 필터링하여 조회되어야 하는 데이터라 생각됩니다..!
또한 예약 가능 상태에 대한 처리 로직이 변경되더라도 데이터 조회 조건은 변경되지 않을 것이라 생각됩니다.

말씀해 주신 부분에 대해 고민해 보고 작성해 보았는데 적절한 대답이었는지 잘 모르겠습니다...!
제가 잘못 생각하고 있는 부분이 있으면 말씀해 주세요!!

f-lab-fi replied:
@puallab 

앗 넵넵 답변 적절합니다.
나중에 DB를 사용한 구현체를 추가했을 때 생각해보시죠. :)

puallab committed:
#9 리뷰반영
1. ScheduleStatus 오탈자 수정
2. Question 클래스 request 의존성 제거
3. InvalidMeetingTimeException 수정
4. QuestionService if문내 예외처리 수정

tilsong reviewed (approved):
👍

puallab committed:
#9 리뷰 반영
ScheduleExcpeion 통합

puallab committed:
Http 상태코드 수정
